text/0097-log-data-model.md

```diff
@@ -0,0 +1,971 @@
+(including multi-line) describing the event in a free form or it can be a
+structured data composed of arrays and maps of other values. Can vary for each
+occurrence of the event coming from the same source.
+
+### Field: `Resource`
+
+Type: key/value pair list.
+
+Description: Describes the source of the log, aka
+[resource](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/overview.md#resources).
+"value" of each pair is of `any` type. Multiple occurrences of events coming
+from the same event source can happen across time and they all have the same
+value of `Resource`. Can contain for example information about the application
+that emits the record or about the infrastructure where the application runs.
+Data formats that represent this data model may be designed in a manner that
+allows the `Resource` field to be recorded only once per batch of log records
+that come from the same source. SHOULD follow OpenTelemetry
+[semantic conventions for Resources](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions).
+
+### Field: `Attributes`
+
+Type: key/value pair list.
+
+Description: Additional information about the specific event occurrence. "value"
+of each pair is of `any` type. Unlike the `Resource` field, which is fixed for a
+particular source, `Attributes` can vary for each occurrence of the event coming
+from the same source. Can contain information about the request context (other
+than TraceId/SpanId). SHOULD follow OpenTelemetry
+[semantic conventions for Attributes](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/trace/semantic_conventions).
+
+## Example Log Records
+
+Below are examples that show one possible representation of log records in JSON.
+These are just examples to help understand the data model. Don’t treat the
+examples as _the_ way to represent this data model in JSON.
+
+This document does not define the actual encoding and format of the log record
+representation. Format definitions will be done in separate OTEPs (e.g the log
+records may be represented as msgpack, JSON, Protocol Buffer messages, etc).
+
+Example 1
+
+```json
+{
+  "Timestamp": 1586960586000, // JSON needs to make a decision about 
+                              // how to represent nanoseconds.
+  "Attributes": {
+    "http.status_code": 500,
+    "http.url": "http://example.com",
+    "my.custom.application.tag": "hello",
+  },
+  "Resource": {
+    "service.name": "donut_shop",
+    "service.version": "semver:2.0.0",
+    "k8s.pod.uid": "1138528c-c36e-11e9-a1a7-42010a800198",
+  },
+  "TraceId": "f4dbb3edd765f620", // this is a byte sequence 
+                                  // (hex-encoded in JSON)
+  "SpanId": "43222c2d51a7abe3",
+  "SeverityText": "INFO",
+  "SeverityNumber": 9,
+  "Body": "20200415T072306-0700 INFO I like donuts"
+}
+```
+
+Example 2
```

Comment:
Here is a sample `body` I crafted with web server access logs as it is an important use case and as it contains many fields.

```
├── body
    ├── http
    │   ├── scheme: https
    │   ├── host: donut.mycie.com
    │   ├── target:/order
    │   ├── method: post
    │   ├── status_code: 500
    │   └── client_ip: 192.0.2.5
    ├── user_agent
    │   ├── name: Chrome
    │   ├── version: 70.0.3538.102
    │   ├── os.name: Mac OS X
    │   ├── os.version: 10.14.0
    │   └── original: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36
```

```

Reply:
Thanks for the example.

Do note that well-defined logs such as web server access logs are expected to follow OpenTelemetry semantic conventions and record the information in the example that you posted in the `Attributes` not in the `Body`.

`Body` is an escape hatch. It is a place to record log data that is perhaps unparsed or does not follow the conventions or otherwise is conceptually different so that it does not fit the fields that the data model defines. `Body` allows to represent any data from another log format as-is, without imposing any particular structure. This is important so that we are able to represent data format and support uses cases that we do not understand today or are not aware of today.

A log Collector should be able to receive forward send log data in protocols that do not necessarily map nicely to this data model. `Body` defined as `any` data type is the escape hatch that makes this possible.

Reply:
Understood. As this may be unexpected to some, I feel it would be valuable to provide such example in the document, I propose to write such access log example. Here is a fragment of the `Attributes` section I worked on, I'll be happy to craft a complete example.

```json
{
   "Attributes":{
      "http.scheme":"https",
      "http.host":"donut.mycie.com",
      "http.target":"/order",
      "http.method":"post",
      "http.status_code":500,
      "http.client_ip":"192.0.2.5",
      "http.referer":"-",
      "http.flavor":"1.1",
      "http.response.bytes":542,
      "user_agent.name":"Chrome",
      "user_agent.version":"70.0.3538.102",
      "user_agent.os.name":"Mac OS X",
      "user_agent.os.version":"10.14.0",
      "user_agent.original":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Safari/537.36",

      "log.file.path": "/var/log/nginx/access.log"
   }
}

```

Reply:
User_agent is supplied as a string from the client, and most web servers aren't going to attempt to parse it. That means that parsing out the user_agent.name, user_agent.version there would need to be a parsing filter somewhere here, which I don't think we currently have in the design, and I'm not sure that we should. Obviously this is a trivial issue with user_agent, but I think it's going to be important to keep separate the structured data that we know when the log is generated and the information that can be extracted from the logged content. I think that's particularly the case when evaluating ECS models, as Elasticsearch is often fed by Logstash, which *does* have explicit parsing and extraction functionality.

Reply:
"http.user_agent" is already defined as a single string in OpenTelemety conventions: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#common-attributes

I added an example to this OTEP.

Reply:
Thanks @tigrannajaryan, I missed this `http.user_agent` field when learning the [OpenTelemetry Semantic Conventions / HTTP](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#common-attributes).

@zenmoto Regarding the parsing of the access log, I see great benefits in:
* parsing the access logs on the web server itself so that it is easy to manage simultaneously the configuration of the access log pattern in the web server config and the configuration of the parsing rule in the log collector. 
* standardising common access logs fields including the decomposition of the user agent to have consistent field name across web servers and load balancers for better search and anomaly detections.

Reply:
@tigrannajaryan two quick questions:
1. Would it be fair to say that it is preferable to use `attributes` and not `body` when possible, so that the log data is better structured and easier to parse?
2. Having the body be of type `any` would make it more difficult for backend storage systems, unless they support some "JSON" type which can be anything.  Restricting it to, say, a list of key-attributes would be helpful for the backend.

Reply:
@cyrille-leclerc - I agree with both the utility of parsing logs, and with the limitation of many log formats. My concern is that log parsing is work, and that it's separate from delivery. If there is a way that we can intercept data before it gets serialized, then that's fantastic- we should, and we should transport it in a structured way. As an example, with Tomcat we can create and inject an object that implements their [AccessLog](https://tomcat.apache.org/tomcat-7.0-doc/api/org/apache/catalina/AccessLog.html) interface, and transport the values in our format.

In other circumstances, our only access to the logs is in an already-written format. In those cases, it's a great deal more work to manage how to extract data from that format, manage the many potential extractions, and decide to know when to apply them. It's not bad work to do, and it does need to be done *somewhere*, but I think it's also a particularly large requirement to back into.

I wanted to call out user_agent specifically because it's a field that requires explicit parsing- it's not something that most web servers understand beyond the string value that was supplied by the browser application. So somewhere a regular expression needs to be applied between where we get the data from the webserver and when it gets serialized into our data model, which implicitly pulls an entire parsing pipeline into scope. 

Personally I think we should work in OpenTelemetry to transport log entries in as fine of a granularity as we can obtain them, but we should explicitly limit our scope to transport, not parsing or refinement. This may mean that for Tomcat we can transfer an access log entry in a parsed and normalized format, but for Apache which doesn't have similar extensibility the entry may need to be transmitted in the Apache Combined Log Format to be parsed and extracted upstream.

Reply:
1. Yes, attributes are more preferable, especially if they are well-known attributes for which we have semantic conventions. And even for custom data it is likely preferable to use attributes since many log backends have nice features around key/value pairs (filtering, searching) but may not have capabilities to work with arbitrary nested `any` type beyond displaying it.

2. `body` is an escape hatch. It needs to support complicated objects. There are log protocols and backends that support such concept. This data model must allow representing it. We should not recommend to generate log records with complicated bodies unless there are good reasons to do so, but for backward compatibility reasons the data model must still be able to support it. If the backend is not capable to represent `any` data in structured form then the particular vendor or logging system has to make a choice about how to flatten this data and encode in a form that it is capable of storing.

Reply:
Resolving. Feel free to reopen if needed.